### ISSUE TITLE
The Explore vs. Exploit Dilemma

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11835]
The Explore vs. Exploit Dilemma<br>
<br>
https://ift.tt/dO78Ppj<br>
<br>
<br>
<br>

<div><div><div><div><div>
<p>I often analogize real-world problems according to their ML-related counterparts. One of them is the exploration-exploitation problem, but it’s often been met with minimal recognition. I wrote this blog as something I can refer to the next time I get a “what do you mean?”</p>
<h2>Introduction: The Multi-Armed Bandit Problem</h2>
<p>Suppose we have a series of decisions to make, each with the potential to yield a reward. In our multi-armed bandit problem, we aim to develop a strategy to maximize this reward over time. We envision each “arm” as a slot machine, each one hiding a different reward distribution. Our task is to identify which arm to pull at each step in time to accumulate the most reward.</p>
<p>If we consider <span><span></span></span> as our starting state—knowing nothing about the reward distributions—and <span><span></span></span> as the ideal state—where we have complete knowledge of the best arm—then we can define a function between ignorance and an optimal selection. In this framework, we can imagine a vector field guiding us from exploring new arms to exploiting the most rewarding ones.</p>
<p>Our state of knowledge at time <span><span></span></span> can be denoted <span><span></span></span>, representing our expected reward for each arm, updated after each trial. We can define the expected reward flow <span><span></span></span> as:</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is an exploration parameter that decreases over time, <span><span></span></span> is the action-value function for each arm (capturing the expected reward given our current understanding), and <span><span></span></span> is a measure of the uncertainty or unexplored potential for each arm.</p>
<p>In early steps, <span><span></span></span>, our strategy should be primarily <strong>exploratory</strong>: <span><span></span></span> and thus <span><span></span></span>. On the contrary, as <span><span></span></span> approaches 1, <span><span></span></span> should approach 0, directing <span><span></span></span> toward the maximum expected value action, or <strong>exploitative</strong> policy. To represent this transition, we can set <span><span></span></span>, where <span><span></span></span> is a decay constant that controls how quickly we shift from exploration to exploitation.</p>
<p>The derivative of this function <span><span></span></span> defines a <em>policy gradient</em>, guiding our choice of action. Instead of learning the policy directly, we train <strong>a forward dynamics model</strong> <span><span></span></span> to predict rewards given our current knowledge, aiming to maximize:</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is the discount factor controlling the impact of future rewards, and <span><span></span></span> is the observed reward at each time step <span><span></span></span>. This function captures our goal to maximize cumulative rewards while balancing exploration and exploitation over time.</p>
<p>The model’s predictions thus guide us to iteratively pull arms that will yield the highest cumulative rewards, refining our understanding of each arm’s distribution and honing in on optimal actions.</p>
<h2>The Forward Dynamics Model</h2>
<p>In the multi-armed bandit problem, a forward dynamics model <span><span></span></span> is an auxiliary model that predicts the expected reward for each arm based on past actions and observed rewards. By approximating the environment’s response to each action, <span><span></span></span> helps us make informed choices, directing our strategy toward higher rewards.</p>
<h4>1. <strong>Defining the Forward Dynamics Model</strong>
</h4>
<p>The forward dynamics model <span><span></span></span> can be structured as a parameterized function with weights <span><span></span></span>, taking in a feature vector <span><span></span></span> (representing each arm’s current state) and outputting a predicted reward <span><span></span></span>. The model aims to approximate the action-value function <span><span></span></span>, which maps each action (arm) to an expected reward. In essence, <span><span></span></span> estimates:</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is the actual reward received from pulling the arm represented by <span><span></span></span>. Training <span><span></span></span> involves refining this approximation over multiple trials, improving its ability to generalize from limited samples.</p>
<h4>2. <strong>Training Objective</strong>
</h4>
<p>To train <span><span></span></span>, we minimize the error between predicted rewards <span><span></span></span> and observed rewards <span><span></span></span>. The model is typically trained by minimizing the mean squared error (MSE) over all arms and trials:</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is the number of training samples, each associated with an action <span><span></span></span> (one of the arms) and an observed reward <span><span></span></span>. Minimizing this loss function <span><span></span></span> encourages <span><span></span></span> to make accurate reward predictions, allowing the model to distinguish more promising arms from less rewarding ones.</p>
<h4>3. <strong>Data Collection Strategy</strong>
</h4>
<p>Training the forward dynamics model requires collecting data on observed rewards from pulling each arm. However, the exploration-exploitation dilemma influences how we gather this data:</p>
<ul>
<li><p><strong>Exploration Phase</strong>: During exploration, the model pulls different arms at random to collect a diverse dataset of reward outcomes. This phase ensures that <span><span></span></span> samples from a wide range of potential rewards, even from suboptimal arms, capturing enough data to model the environment’s variability.</p></li>
<li><p><strong>Exploitation Phase</strong>: As the model shifts toward exploitation, it begins to pull the arms predicted to yield higher rewards based on <span><span></span></span>’s predictions. During this phase, <span><span></span></span> refines its understanding of the best arms and focuses on modeling the nuances in expected rewards for these higher-value actions.</p></li>
</ul>
<h4>4. <strong>Incorporating Reward Predictions into Policy Gradients</strong>
</h4>
<p>Once trained, <span><span></span></span>’s reward predictions are used to inform the policy and guide our decision-making. The policy gradient objective, which maximizes cumulative rewards, can be modified to include predicted rewards:</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is the discount factor for future rewards, <span><span></span></span> is the action-state at time <span><span></span></span>, and <span><span></span></span> provides the reward prediction. This objective encourages the model to pull arms with higher predicted rewards, continually adjusting based on new data from <span><span></span></span>.</p>
<p>Ultimately, the forward dynamics model <span><span></span></span> enables a structured approach to decision-making by predicting rewards for each arm, balancing exploration and exploitation based on estimated reward distributions. With ongoing training, <span><span></span></span> adapts to new information, ensuring the model’s predictions remain accurate as we refine our understanding of each arm. This iterative learning process helps maximize cumulative rewards by aligning actions with <span><span></span></span>’s growing knowledge of the environment.</p>
<h2>The Exploration-Exploitation Dilemma</h2>
<p>The exploration-exploitation dilemma is central to multi-armed bandit problems: we need to <em>explore</em> enough to understand each arm’s potential while also <em>exploiting</em> known information to maximize immediate rewards. The parameter <span><span></span></span>, which governs the rate at which we shift from exploration to exploitation, plays a key role in this balance.</p>
<p>Choosing <span><span></span></span> is a careful process. If <span><span></span></span> is too large, <span><span></span></span> decreases rapidly, leading the model to quickly favor exploitation. In this case, we may prematurely commit to arms that appear optimal based on limited early data, potentially missing out on higher-reward options that remain unexplored. On the other hand, if <span><span></span></span> is too small, <span><span></span></span> remains high, and we may spend too much time exploring, failing to capitalize on accumulated knowledge to maximize rewards.</p>
<p>To choose an optimal <span><span></span></span>, consider the following factors:</p>
<ol>
<li><p><strong>Expected Variance Across Arms</strong>: If the rewards from each arm vary significantly, then a smaller <span><span></span></span> is generally favorable because it allows more exploration. This ensures that our policy gathers enough information to accurately assess the best arm. On the contrary, if we expect rewards to be fairly consistent across arms, we may prefer a larger <span><span></span></span>, quickly shifting to exploitation since exploration is less likely to yield drastically different results.</p></li>
<li><p><strong>Risk Tolerance</strong>: If our strategy prioritizes high-risk, high-reward outcomes, a smaller <span><span></span></span> allows more exploration, potentially discovering arms with rare but substantial rewards. Conversely, if the strategy is risk-averse, a larger <span><span></span></span> may be preferable, reducing the time spent on uncertain options and focusing on the most reliable rewards found early on.</p></li>
<li><p><strong>Adaptive <span><span></span></span> Adjustments</strong>: In complex environments, an adaptive approach to <span><span></span></span> can be beneficial. Here, <span><span></span></span> might dynamically adjust based on observed reward distributions or changes in the variance of <span><span></span></span>. For instance, if new arms produce high variance in <span><span></span></span>, <span><span></span></span> could decrease temporarily to encourage exploration; once variance stabilizes, <span><span></span></span> can increase to favor exploitation.</p></li>
</ol>
<p>There are also added considerations, such as the total number of trials <span><span></span></span>and the decay function used to set <span><span></span></span>. Depending on how critical initial exploration is or how smooth one wants to transition between exploration and exploitation, we may choose a different decay function, such as exponential or reciprocal. If <span><span></span></span> is large, we can afford to spend more time exploring, and thus a smaller <span><span></span></span> is suitable. However, if <span><span></span></span> is limited, we should increase <span><span></span></span> to favor quicker convergence toward exploitation.</p>
<p>Overall, the goal is to maximize cumulative reward over time by balancing exploration and exploitation. The parameter <span><span></span></span> determines this balance by controlling the rate of decay in the exploration factor <span><span></span></span>. Experimenting with different values of <span><span></span></span> and decay functions is often necessary to find a balance that aligns with the specific requirements of the problem.</p>
<p>By fine-tuning <span><span></span></span> and <span><span></span></span>, we guide the model toward the most rewarding actions while avoiding premature convergence on suboptimal options.</p>
<h2>The Prolonged Analogy</h2>
<p>I find the many people I know go through the exploration-exploitation dilemma. No one wants to be Esther Greenwood, but neither does anyone want to dive completely into some repetitive FANG lifestyle, living somebody else’s dream before they’ve even discovered their own. How do you choose when to explore and when to exploit?</p>
<p>Surely, we should have an adaptive <span><span></span></span> that depends on our surroundings. How does your <span><span></span></span> change as all the people around you found B2B SaaS startups (no shade, simply not my kind of intellectual stimulation), go into finance, and marry college lovers? Can I bet on this arm and go all in, or should I hedge my bets and let my forward dynamics model better-model the environmental variability?</p>
<p>Uncertainty can be modeled as</p>
<p><span><span><span></span></span></span></p>
<p>where <span><span></span></span> is the number of ensemble models and <span><span></span></span> is the reward prediction from forward dynamics model <span><span></span></span>. We unfortunately do not have the benefit of an ensemble, but we can utilize people who are close to us (and hopefully similar in thought process) to provide a proxy for uncertainty.</p>
<p>I have done a lot of exploring all throughout high school, studying many different scientific subjects for various olympiad competitions, working in various research groups, and having a general exposure to what the final <span><span></span></span> exploitative policy would look like for many possible arms, whether SWE, quant, or academia. Additionally, I feel that exploration is a part of my reward function. An entropy parameter.</p>
<p>So, culminating everything, my dream would be industry research or founding a successful research-heavy startup. You don’t just have a sizeable salary and prestiege to move into any other field you are interested in, but also are continually growing your personal capital through learning/exploration. I also have less uncertainty than most, given that most people I know believe I can do industry research or tackle difficult deeptech problems within a startup. Above all, I am building towards a relatively high risk tolerance — I am okay if a startup fails (it won’t). Regardless, I have a pretty long-range <span><span></span></span> and life offers a relatively high expected variance across its arms.</p>
<p>I was going to make up a decay function for my <span><span></span></span>, but the truth is, it’s probably something like</p>
<p><span><span><span></span></span></span></p>
<p>Thanks for reading!</p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Hello <br>
<br>
October 14, 2024 at 03:37PM
